libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,995 @@
+
+		vector<SourceLocation> output;
+		if (auto decl = sourceIdentifier->annotation().referencedDeclaration)
+			output += findAllReferences(decl, decl->name(), sourceUnit);
+
+		for (auto const decl: sourceIdentifier->annotation().candidateDeclarations)
+			output += findAllReferences(decl, decl->name(), sourceUnit);
+		return output;
+	}
+	else if (auto const decl = dynamic_cast<VariableDeclaration const*>(sourceNode))
+	{
+		auto const sourceName = _documentPosition.path;
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		return findAllReferences(decl, decl->name(), sourceUnit);
+	}
+	else if (auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		return findAllReferences(functionDefinition, functionDefinition->name(), sourceUnit);
+	}
+	else if (auto const* enumDef = dynamic_cast<EnumDefinition const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		return findAllReferences(enumDef, enumDef->name(), sourceUnit);
+	}
+	else if (auto const memberAccess = dynamic_cast<MemberAccess const*>(sourceNode))
+	{
+		if (Declaration const* decl = memberAccess->annotation().referencedDeclaration)
+		{
+			auto const sourceName = _documentPosition.path;
+			frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+			return findAllReferences(decl, memberAccess->memberName(), sourceUnit);
+		}
+		return {};
+	}
+	else if (auto const* importDef = dynamic_cast<ImportDirective const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		return findAllReferences(importDef, importDef->name(), sourceUnit);
+	}
+
+	return {};
+}
+
+vector<DocumentHighlight> LanguageServer::semanticHighlight(DocumentPosition _documentPosition)
+{
+	if (!m_compilerStack || m_compilerStack->state() < frontend::CompilerStack::AnalysisPerformed)
+		return {};
+
+	frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(_documentPosition.path);
+	ASTNode const* sourceNode = findASTNode(_documentPosition.position, _documentPosition.path);
+	if (!sourceNode)
+		return {};
+
+	vector<DocumentHighlight> output;
+
+	matchASTNode(
+		sourceNode,
+		[&](Identifier const* sourceIdentifier)
+		{
+			if (sourceIdentifier->annotation().referencedDeclaration)
+				output += ReferenceCollector::collect(sourceIdentifier->annotation().referencedDeclaration, sourceUnit, sourceIdentifier->name());
+
+			for (Declaration const* declaration: sourceIdentifier->annotation().candidateDeclarations)
+				output += ReferenceCollector::collect(declaration, sourceUnit, sourceIdentifier->name());
+
+			for (Declaration const* declaration: sourceIdentifier->annotation().overloadedDeclarations)
+				output += ReferenceCollector::collect(declaration, sourceUnit, sourceIdentifier->name());
+		},
+		[&](VariableDeclaration const* varDecl)
```

Comment:
If you use `dynamic_cast` to declaration, you can combine all these cases, and that's what the inheritance hierarchy is for